### PR TITLE
Add tests for call.cancel() from within messageRead

### LINKED ISSUE #1682: Cancelling a call results in onClose never being invoked.
Consider the following code:

```
      final ClientCall<Void, Void> call = channel.newCall(..., ...);
      ClientCall.Listener<Void> listen =
          new ClientCall.Listener<Void>() {
            @Override
            public void onMessage(Void message) {
              call.cancel();
            }

            @Override
            public void onClose(Status status, Metadata trailers) {
              // NEVER CALLED!
            }
          };
      call.start(listen, new Metadata());
      call.sendMessage(msg);
      call.halfClose();
      call.request(1);
```

onClose will not be invoked, so it becomes impossible to know when the call is actually done.  This seems wrong.  If the call is cancelled from outside of the listener, onClose is invoked.

### PR DESCRIPTION
**Pull request by buchgr**
I can't reproduce the bug reported in #1682 ... can you confirm? Am I missing something?

### CODE REVIEW

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java [hunk 1]
```diff
@@ -587,6 +589,63 @@ public void clientCancel() throws Exception {
...
+    ServerStreamListener mockServerStreamListener = serverStreamCreation.listener;
+    verify(mockServerStreamListener, timeout(TIMEOUT_MS)).onReady();
+
+    assertTrue(serverStream.isReady());
+    serverStream.writeHeaders(new Metadata());
+    serverStream.writeMessage(methodDescriptor.streamRequest("foo"));
+    serverStream.flush();
+
+    while (!closedCalled.get()) {
+      LockSupport.parkNanos(TimeUnit.MILLISECONDS.toNanos(100));
```
Comment (carl-mastrangelo) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: LockSupport is really intended for building locking constructs.  This code would be simpler if you just used SettableFuture above, and hung on get() here.

**Comment by buchgr** [comment]
@carl-mastrangelo PTAL

**Comment by carl-mastrangelo** [comment]
@buchgr 

I have a repro here:

https://github.com/carl-mastrangelo/grpc-java/commit/786a96f6bf14035ca81849285ad5a65a4fe2dd91


Run NoiseClient and it will eventually stop responding, usually within 1 or 2 iterations.

**Comment by carl-mastrangelo** [comment]
Actually, the original bug report may be incorrect.  It actually is seeming like onHeaders is not being called.  I can confirm the headers are sent to the server, but the response never seems to make it back.

**Comment by buchgr** [comment]
@carl-mastrangelo ahh makes sense :-).

Probably still useful to add these additional tests. Sounds good?

**Comment by carl-mastrangelo** [comment]
@buchgr LGTM

**Comment by buchgr** [comment]
@carl-mastrangelo thanks!

**Pull request merged**